core/shared/src/main/scala/zio/RefM.scala

```diff
@@ -146,7 +146,7 @@ object RefM extends Serializable {
       interrupted.get.flatMap {
         case Some(cause) => onDefect(cause)
         case None =>
-          update(a).foldM(e => onDefect(Cause.fail(e)) <* promise.fail(e), {
+          update(a).foldCauseM(c => onDefect(c).ensuring(promise.halt(c)), {
```

Comment:
Can you add a test for this change?

Reply:
Of course, I will look into this over the weekend.

Reply:
I've just added a test (see fecf6b412ac00fb7afe3b8c7bb23067656500eb5). Without the change from above, this test times out.